DiceKit/FrequencyDistribution.swift

```diff
@@ -0,0 +1,149 @@
+
+public func == (lhs: FrequencyDistribution, rhs: FrequencyDistribution) -> Bool {
+    return lhs.outcomesPerValue == rhs.outcomesPerValue
+}
+
+// MARK: - Operators
+
+extension FrequencyDistribution {
+    
+    // MARK: Primitive Operations
+    
+    public func negateValues() -> FrequencyDistribution {
+        let newOutcomesPerValue = outcomesPerValue.mapKeys {
+            (value, _) in -value
+        }
+        
+        return FrequencyDistribution(newOutcomesPerValue)
+    }
+    
+    public func shiftValues(value: Value) -> FrequencyDistribution {
+        let newOutcomesPerValue = outcomesPerValue.mapKeys {
+            (baseValue, _) in baseValue + value
+        }
+        
+        return FrequencyDistribution(newOutcomesPerValue)
+    }
+    
+    public func scaleOutcomes(scalar: Double) -> FrequencyDistribution {
+        let newOutcomesPerValue = outcomesPerValue.mapValues {
+            (_, outcome) in outcome * scalar
+        }
+        
+        return FrequencyDistribution(newOutcomesPerValue)
+    }
+    
+    public func normalizeOutcomes() -> FrequencyDistribution {
+        let outcomes = totalOutcomes
+        let normalizedOutcomesPerValue = outcomesPerValue.mapValues {
+            (_, outcome) in outcome / outcomes
+        }
+        
+        return FrequencyDistribution(normalizedOutcomesPerValue)
+    }
+    
+    // MARK: Advanced Operations
+    
+    public func add(x: FrequencyDistribution) -> FrequencyDistribution {
+        var newOutcomesPerValue = outcomesPerValue
+        for (value, outcome) in x.outcomesPerValue {
+            let exisitingOutcome = newOutcomesPerValue[value] ?? 0
+            let newOutcome = exisitingOutcome + outcome
+            newOutcomesPerValue[value] = newOutcome
+        }
+        
+        return FrequencyDistribution(newOutcomesPerValue)
+    }
+    
+    // TODO: public func subtract(x: FrequencyDistribution) -> FrequencyDistribution
+    // Probably needed to support divide
+    
+    public func multiply(x: FrequencyDistribution) -> FrequencyDistribution {
+        return outcomesPerValue.reduce(.additiveIdentity) {
+            let (value, outcome) = $1
+            let addend = x.shiftValues(value).scaleOutcomes(outcome)
+            return $0.add(addend)
+        }
+    }
+    
+    // TODO: public func divide(x: FrequencyDistribution) -> FrequencyDistribution
+    // Needed to support power(-n)
+    
+    public func power(x: Int) -> FrequencyDistribution {
+        guard x != 0 else { return .multiplicativeIdentity }
+        
+        // Crappy implementation. Currently O(n). Can be O(log(n)).
+        // TODO: Handle negative properly
+        var freqDist = self
+        for _ in 1..<abs(x) {
+            freqDist = freqDist.multiply(self)
+        }
+        
+        return freqDist
+    }
+    
+    // This needs a better name, it's basically n*x^y
+    public func powerMultiply(x: FrequencyDistribution) -> FrequencyDistribution {
```

Comment:
Can't think of a super intuitive name.. The symbolic expression being in the comment does help a lot though.